include/dlaf/sender/transform_mpi.h

```diff
@@ -0,0 +1,179 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <pika/unwrap.hpp>
+
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/sender/transform.h"
+#include "dlaf/sender/when_all_lift.h"
+
+namespace dlaf::comm::internal {
+
+/// Helper for moving a PromiseGuard<Communicator> into the MPI function being
+/// called by transformMPI. Callables passed to transformMPI have their
+/// arguments passed by reference, but doing so with PromiseGuard would keep the
+/// guard alive until the completion of the MPI operation, whereas we are only
+/// looking to guard the submission of the MPI operation. We therefore use this
+/// wrapper to move PromiseGuard<Communicator> into a transformMPI callables so
+/// that the guard is released on return from the callable.
+template <typename T>
+decltype(auto) movePromiseGuard(T& t) {
+  if constexpr (std::is_same_v<std::decay_t<T>, dlaf::common::PromiseGuard<Communicator>>) {
+    return std::move(t);
+  }
+  else {
+    return static_cast<T&>(t);
+  }
+}
+
+/// Helper type for wrapping MPI calls.
+///
+/// Wrapper type around calls to MPI functions. Provides a call operator that
+/// creates an MPI request and passes it as the last argument to the provided
+/// callable. The wrapper then waits for the the request to complete with
+/// yield_while.
+///
+/// This could in theory be a lambda inside transformMPI.  However, clang at
+/// least until version 12 fails with an internal compiler error with a trailing
+/// decltype for SFINAE. GCC has no problems with a lambda.
```

Comment:
FYI, this seems to be a close to minimal reproducible example of the failure: https://github.com/pika-org/pika/issues/192.